[optim] Allow torch.float64 scalars for forloop + foreach implementations

janeyx99 opened the pull request:
Should allow for uses cases mentioned in #110940

This would allow scalars to also be float64s in the foreach implementation. The fused implementation would still create a float32 step on Adam and AdamW. This PR also does NOT worry about performance and is mainly for enablement.

Next steps:
- Relax the constraint on fused adam(w) and allow torch.float64 scalars there
- Allow _performant_ mixed dtypes in foreach (a bigger project in itself).

This PR will conflict with my other PRs, I will figure out a landing order

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115841

janeyx99 committed:
Allow torch.float64 scalars for foreach implementations

[ghstack-poisoned]

albanD commented on test/optim/test_optim.py:
```diff
@@ -770,11 +770,10 @@ def _test_derived_optimizers(self, optimizer_pairs_with_flags, flag, reduced_pre
+                empty_param = torch.empty((), device=device, dtype=torch.float64)
+                parameters = list(model.parameters()) + [empty_param]
 
-                optimizer = optimizer_constructor(
-                    list(model.parameters()) + empty_params, **params_with_flags
-                )
+                optimizer = optimizer_constructor(parameters, **params_with_flags)
 
```
This is just refactoring right?

janeyx99 replied:
This allows us to access the step of the first param later :p

albanD commented on torch/optim/optimizer.py:
```diff
@@ -132,6 +132,11 @@ def _view_as_real(params, *state_and_grads):
+def _get_scalar_dtype(is_fused=None):
+    if is_fused:
+        return torch.float32
+    return torch.float64 if torch.get_default_dtype() == torch.float64 else torch.float32
```
I don't remember the story for default dtype vs default Tensor type. Do we need to check both here?

janeyx99 replied:
I don't think we do, get_default_dtype seems to work as expected:
<img width="764" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/be395f5a-69f4-495c-8944-e49b234d8dde">

janeyx99 committed:
Update on "[optim] Allow torch.float64 scalars for forloop + foreach implementations"

Should allow for uses cases mentioned in #110940

This would allow scalars to also be float64s in the foreach implementation. The fused implementation would still create a float32 step on Adam and AdamW. This PR also does NOT worry about performance and is mainly for enablement.

Next steps:
- Relax the constraint on fused adam(w) and allow torch.float64 scalars there
- Allow _performant_ mixed dtypes in foreach (a bigger project in itself).

This PR will conflict with my other PRs, I will figure out a landing order




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

janeyx99 committed:
Update on "[optim] Allow torch.float64 scalars for forloop + foreach implementations"

Should allow for uses cases mentioned in #110940

This would allow scalars to also be float64s in the foreach implementation. The fused implementation would still create a float32 step on Adam and AdamW. This PR also does NOT worry about performance and is mainly for enablement.

Next steps:
- Relax the constraint on fused adam(w) and allow torch.float64 scalars there
- Allow _performant_ mixed dtypes in foreach (a bigger project in itself).

This PR will conflict with my other PRs, I will figure out a landing order




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
